DataFrames review (#69)
JuliaDataScience/JuliaDataScience: contents/dataframes_indexing.md

""")
```

However, converting a `DataFrame` to a `Dict` is only useful when the elements are unique.
Generally that is not the case, it is time to learn how to `filter` a `DataFrame`.

## Filter and Subset {#sec:filter_subset}

There are two ways to remove rows from a `DataFrame`, one is `filter` (@sec:filter) and the other is `subset` (@sec:subset).
`filter` was added earlier to `DataFrames.jl`, is more powerful, and more consistent with syntax from Julia base, so that is why we start with discussing `filter`.
`subset` is newer and often more convenient.

### Filter {#sec:filter}

From this point on, we start to get into the more powerful features of `DataFrames.jl`.

> To do this, we need to learn some functions, such as `select` and `filter`.

Reviewer: I copied the word "verbs" from bkamins: https://discourse.julialang.org/t/pull-dataframes-columns-to-the-front/60327/5.

I agree though that the word "functions" is clearer, but could you doublecheck by looking at the discourse post?

Author: Added complexity... Verbs and Functions here can be exchangeable.

I don't think we would benefit to strictly use the word verbs and then having to define what a verb is to the reader.

But don't worry!
It might be a relieve to know that the **general design goal of `DataFrames.jl` is to keep the number of functions that a user has to learn to a minimum[^verbs]**.

[^verbs]: According to Bogumił Kamiński (lead developer and maintainer of `DataFrames.jl`) on [Discourse](https://discourse.julialang.org/t/pull-dataframes-columns-to-the-front/60327/5).

Like before, we resume from the `grades_2020`:

```jl
sco("grades_2020()"; process=without_caption_label)
```

We can filter rows by using `filter(source => f::Function, df)`.
Note how this function is very similar to the function `filter(f::Function, V::Vector)` from Julia `Base` module.
This is because `DataFrames.jl` uses **multiple dispatch** (see @sec:multiple_dispatch) to define a new method of `filter` that accepts a `DataFrame` as argument.

At first sight, defining and working with a function `f` for filtering can be a bit difficult to use in practice.

> Hold tight, that the effort is well-paid, since **it is very powerful way of filtering data**.

Reviewer: You're definitely better at writing in an exciting way :+1: :+1:

"""; process=without_caption_label)
```

To get all the rows which are *not* Alice, `==` (equality) can be replaced by `!=` (inequality) in all previous examples:

```jl
s = """filter(:name => !=("Alice"), grades_2020())"""
sco(s; process=without_caption_label)
```

> Now, to show **why anonymous functions are so powerful**, we can come up with a slight more complex filter.

Reviewer: ```suggestion
Now, to show **why anonymous functions are so powerful**, we can come up with a slightly more complex filter.
```

> In this filter, we want to have the people whose name start with A or B **and** while also having grades over 6:

Reviewer: ```suggestion
In this filter, we want to have the people whose name start with A or B **and**  have a grade above 6:
```
Not like this? The "while" sounds off: "Select the people where A and B" versus "Select the people where A and while B".

### Subset {#sec:subset}

The `subset` function was added to make it easier to work with missing values (@sec:missing_data).
In contrast to `filter`, `subset` works on complete columns instead of rows or single values.
If we want to use our earlier defined functions, we should wrap it inside `ByRow`:

```jl
s = "subset(grades_2020(), :name => ByRow(equals_alice))"
sco(s; process=without_caption_label)
```

Also note that the `DataFrame` is now the first argument `subset(df, args...)`, whereas in `filter` it was the second argument `filter(f, df)`.

> The reason for this is that Julia defines filter as `filter(f, V::Vector)` and `DataFrames.jl` chose to maintain consistency with existing Julia functions that were extended to `DataFrame`s types by multiple dispatch.

Reviewer: nice!

@sco salaries()
```

This data is about a plausible situation where you want to figure out your colleagues' salaries, and haven't figured it out for Zed yet.
Even though we don't want to encourage these practices, we suspect it is an interesting example.

> Suppose we want to know who earns more than 2000.

Author: Or dollars?

sco(s; process=without_caption_label)
```

```{=comment}

> Rik, we need a example of both filter and subset with multiple conditions, as in:

Reviewer: That sounds okay. Can we move this comment into an issue? You can assign me to it. I think that TODOs inside code are very difficult to track.

Author: updated #54
